### PR TITLE
Get list of project to check from RoslynProjectProvider

### PR DESCRIPTION
**Pull request by duncanp-sonar**
- if there are no projects then there is nothing to clean

Part of #4386

### CODE REVIEW

File: src/ConnectedMode/Migration/RoslynProjectProvider.cs [hunk 1]
```diff
@@ -42,6 +42,7 @@ internal class RoslynProjectProvider : IRoslynProjectProvider
 
         [ImportingConstructor]
         public RoslynProjectProvider(VisualStudioWorkspace vsWorkspace)
+            : this(new VisualStudioWorkspaceWrapper(vsWorkspace))
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/RoslynProjectProvider.cs:45]: Oopsie - minor problem with the wrapper that I found when debugging...

File: src/ConnectedMode.UnitTests/Migration/FileProviders/XmlFileProviderTests.cs [hunk 2]
```diff
@@ -177,5 +234,42 @@ private static Mock<IVsSolution> CreateIVsSolution(string pathToReturn)
...
+        {
+            var projectProvider = new Mock<IRoslynProjectProvider>();
+
+            var projects = CreateProjects(projectFilePaths);
+            projectProvider.Setup(x => x.Get()).Returns(projects);
+
+            return projectProvider;
+        }
+
+        private static IReadOnlyList<Project> CreateProjects(params string[] projectFilePaths)
```
Comment (duncanp-sonar) [inline src/ConnectedMode.UnitTests/Migration/FileProviders/XmlFileProviderTests.cs:248]: These helper methods aren't the same as the ones I wrote for the `RoslynProjectProvider` itself i.e. we can share them.

These tests need to mock the project file path only.

File: src/ConnectedMode/Migration/FileProviders/XmlFileProvider.cs [hunk 3]
```diff
@@ -39,16 +39,19 @@ namespace SonarLint.VisualStudio.ConnectedMode.Migration.FileProviders
...
         private readonly IServiceProvider serviceProvider;
+        private readonly IRoslynProjectProvider projectProvider;
         private readonly ILogger logger;
         private readonly IThreadHandling threadHandling;
         private readonly IFileSystem fileSystem;
 
+        // We only search for non-project files - we get the Roslyn project file paths
+        // from the VSWorkspace.
+        // Corner case: this means that we won't clean projects that are in the solution
+        // but unloaded, since they won't appear in the VsWorkspace
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/FileProviders/XmlFileProvider.cs:50]: That way, there's no risk that we pick up any projects that are not part of the solution.

Also, it means we can skip looking for other types of file to clean up: no Roslyn projects = no files to check for settings.

File: src/ConnectedMode/Migration/FileProviders/XmlFileProvider.cs [hunk 4]
```diff
@@ -115,14 +143,14 @@ private IEnumerable<string> FindFiles(string rootFolder)
 
             var timer = Stopwatch.StartNew();
 
-            var allMatches = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            var allMatches = new List<string>();
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/FileProviders/XmlFileProvider.cs:146]: I moved the de-duplication using a hashset to the top-level, so we can just use a List here and call List.AddRange.

**Comment by duncanp-sonar** [comment]
I've manually tested the behaviour with Roslyn and no-Roslyn projects. In the no-Roslyn project case, the reset of the migration process still happens. 
We could update progress messages to make it clearer is nothing to clean in that case, although we do log this in the output window.

File: src/ConnectedMode/Migration/RoslynProjectProvider.cs (hunk 1 above)
Comment (bigfluffycookie) [reply src/ConnectedMode/Migration/RoslynProjectProvider.cs:45]: Ah, yea also missed that

**Review by bigfluffycookie — APPROVED** [review]

**Pull request merged**